libs/runtime/ebpf_object.c

```diff
@@ -509,83 +477,64 @@ ebpf_object_reference_by_handle(
-
-    uint32_t index;
-    ebpf_result_t result = _get_index_from_id(id, &index);
+    ebpf_id_entry_t* entry = NULL;
+    ebpf_result_t result = ebpf_hash_table_find(_ebpf_id_table, (const uint8_t*)&id, (uint8_t**)&entry);
     if (result != EBPF_SUCCESS) {
         goto Done;
     }
 
-    ebpf_id_entry_t* entry = &_ebpf_id_table[index];
-    if (entry->reference_count <= 0) {
-        __fastfail(FAST_FAIL_INVALID_REFERENCE_COUNT);
-    }
-
-    // ref count is non-zero
+    // Skip entries that have been deleted.
     if (entry->object == NULL) {
-
-        // The object at this entry has been deleted and all that remains are (the now stale) references to this entry
-        // held by other objects.  This is a non-reversible situation and there's no point in giving out references
-        // for such entries, so deny with a suitable error code.
-        result = EBPF_STALE_ID;
+        result = EBPF_INVALID_OBJECT;
         goto Done;
     }
 
-    ebpf_assert(entry->object->type == object_type);
-    if (entry->object->type != object_type) {
+    // Skip entries that are not of the requested type.
+    if (entry->type != object_type) {
         result = EBPF_INVALID_OBJECT;
         goto Done;
     }
 
-    // We have a live object of the matching type and with an existing non-zero ref count, so we're good.
-    // We're under lock so un-protected access is ok.
-    entry->reference_count++;
-    result = EBPF_SUCCESS;
-    _update_reference_history(entry, EBPF_OBJECT_ACQUIRE, file_id, line);
+    // Acquire a reference on the entry.
+    int64_t new_refcount = ebpf_interlocked_increment_int64(&entry->reference_count);
+    if (new_refcount <= 0) {
+        __fastfail(FAST_FAIL_INVALID_REFERENCE_COUNT);
+    }
+
+    // Update the reference history.
+    ebpf_object_update_reference_history(entry, EBPF_OBJECT_ACQUIRE, file_id, line);
 
 Done:
-    ebpf_lock_unlock(&_ebpf_object_tracking_list_lock, state);
     return result;
 }
 
-_Must_inspect_result_ ebpf_result_t
+void
 ebpf_object_release_id_reference(ebpf_id_t id, ebpf_object_type_t object_type, uint32_t file_id, uint32_t line)
 {
-    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
-
-    uint32_t index;
-    ebpf_result_t result = _get_index_from_id(id, &index);
+    ebpf_id_entry_t* entry = NULL;
+    // Find the entry in the ID table.
+    ebpf_result_t result = ebpf_hash_table_find(_ebpf_id_table, (const uint8_t*)&id, (uint8_t**)&entry);
     if (result != EBPF_SUCCESS) {
-        goto Done;
+        // This should never happen as it means the caller is trying to release a reference
+        // to an entry that does not exist.
+        __fastfail(FAST_FAIL_INVALID_REFERENCE_COUNT);
     }
 
-    ebpf_id_entry_t* entry = &_ebpf_id_table[index];
-    if (entry->reference_count <= 0) {
+    if (entry->type != object_type) {
         __fastfail(FAST_FAIL_INVALID_REFERENCE_COUNT);
```

Comment:
Should this be `FAST_FAIL_OBJECT_TYPE_MISMATCH` or something similar?

Reply:
If it's the wrong type, then something has gone terribly wrong, so just terminate.